### PR TITLE
Add documentation for the LPAD and RPAD OQL functions

### PR DESCRIPTION
**Pull request by passalaqua**
~Should be added in 11.11.0~

Changed to 11.12.0 in a196c91a4d95b04b39d1723f291996c8a9dbff5f

### CODE REVIEW

**Comment by katarzyna-koltun-mx** [comment]
Awaiting 11.11 release.

**Comment by passalaqua** [comment]
> Awaiting 11.11 release.

Just FYI, this still requires approval from the team and for the implementation story to be finalised.

**Comment by katarzyna-koltun-mx** [comment]
Got it, will wait for the go ahead.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 1]
```diff
@@ -1329,6 +1331,57 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+LPAD ( expression , length_expression [, pad_expression ] )
+```
+
+#### expression
+
+`expression` specifies the expression of type `string` to pad.
+
+#### length_expression
+
+`length_expression` specifies, as an `integer`, the length of the resulting string.
```
Comment (pijuskri) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: I would propose a different wording:

> `length_expression` specifies the length of the resulting string. The expression must be of type `integer` or `long`

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 2]
```diff
@@ -1329,6 +1331,57 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+
+{{% alert color="info" %}}
+If `length_expression` is smaller than the length of `expression`, this function truncates it. This behavior is database specific.
+{{% /alert %}}
+
+#### pad_expression
+
+`pad_expression` is an optional parameter that specifies the character or string to pad with. If not specified, the space character is used.
```
Comment (pijuskri) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1370]: Let's visualize the character:

> the space character (" ") is used

**Review by pijuskri — CHANGES_REQUESTED** [review]

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 2 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1370]: I went with this description to be consistent with the documentation for TRIM and, imo, adding such a visualization is always a bit confusing. If I were to change here, I think it's better to change in the rest of the page as well.

Comment (pijuskri) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1370]: I don't find it to be a big deal either way, so lets keep it as is. Other parts of the documentation that refers to spaces also dont show `(" ")`

**Comment by pijuskri** [comment]
Approved by team for technical review

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 3]
```diff
@@ -1329,6 +1331,57 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+
+#### Syntax
+
+```sql
+LPAD ( expression , length_expression [, pad_expression ] )
+```
+
+#### expression
+
+`expression` specifies the expression of type `string` to pad.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1355]: Any special behavior for `NULL`? For Trim it returns `NULL` but does this just return a whole load of `pad_expression`?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 4]
```diff
@@ -1329,6 +1331,57 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+
+#### length_expression
+
+`length_expression` specifies the length of the resulting string. The expression must be of type `integer` or `long`
+
+{{% alert color="info" %}}
+If `length_expression` is smaller than the length of `expression`, this function truncates it. This behavior is database specific.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1365]: Truncates the right side, even though we are doing `LPAD`?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 5]
```diff
@@ -1329,6 +1331,57 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+
+
+```sql
+SELECT LPAD('hello', 10, 'x') AS padded FROM Sales.Order
+```
+
+| padded     |
+|:-----------|
+| xxxxxhello |
+
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1393]: Is it worth an example with a string `pad_expression`?

```sql
SELECT LPAD('hello', 10, 'abc') AS padded FROM Sales.Order
```

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 6]
```diff
@@ -1507,6 +1560,57 @@ SELECT ROUND((Price : 7), 2) as RoundedPrice, Price : 7 FROM Sales.Order
...
+#### expression
+
+`expression` specifies the expression of type `string` to pad.
+
+#### length_expression
+
+`length_expression` specifies the length of the resulting string. The expression must be of type `integer` or `long`
+
+{{% alert color="info" %}}
+If `length_expression` is smaller than the length of `expression`, this function truncates it.
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1649]: Still normal truncation - so the same as LPAD?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 7]
```diff
@@ -1507,6 +1560,57 @@ SELECT ROUND((Price : 7), 2) as RoundedPrice, Price : 7 FROM Sales.Order
 | 0.33         | 3.33333333 |
 | 1.17         | 1.17142857 |
 
+### RPAD {#rpad-function}
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1621]: Same comments as for LPAD

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md [hunk 8]
```diff
@@ -1329,6 +1331,57 @@ SELECT * FROM Sales.Customer WHERE LOWER(LastName) = 'doe'
...
+
+#### Examples
+
+```sql
+SELECT LPAD('hello', 10) AS padded FROM Sales.Order
+```
+
+| padded     |
+|:-----------|
+|      hello |
```
Comment (MarkvanMents) [inline content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: This looks wrong when built. May have to think around this (`&nbsp;` would work - but still difficult to see that it is space padded rather than right justified)

Image

)

**Review by MarkvanMents — COMMENTED** [review]
Thanks for the PR. I have a few questions:

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 3 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1355]: They just return `null` if either the expression or length are null. If the padding pattern is `null`, it is usually null (HSQLDB is the exception and throws an error).

And if `expression` is the empty string or the length is zero, then Oracle returns null, but other databases return the empty string.

And if the padding pattern is the empty string, then Oracle and MariaDB return null, HSQLDB uses the space character, MySQL returns the empty string, and other databases return the original string. It's kind of a mess and I suppose those are cases where things can be described as "it's database dependent", so I didn't add to this page.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 4 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1365]: Weirdly, yes. We were just as confused when implementing. The only exception is HSQLDB, which does the right thing, but possibly by accident given how special it is.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 5 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1393]: Probably a good idea. I can write it as part of the same query (shorter page) or as a separate example (clearer). Which one do you prefer?

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 6 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1649]: Yep.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 8 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: I suppose we can change the query to

`SELECT '|' + LPAD('hello', 10) + '|' AS padded FROM Sales.Order`

The resulting string would be `|     hello|`. Would we render it correctly with spaces?

Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: We could also just render the result as

```
·····hello
```

Or something similar and indicate `·` represents a whitespace.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 3 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1355]: Is it worth noting that this edge case behavior depends on the database and that the customer should test them on their configuration? I'm not sure that all our users would think about this, or might raise questions about it.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 4 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1365]: Thanks for the answer. We've said that it's database specific, so let's just leave it like that, rather than trying to give an actual answer.

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 8 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md]: Good idea. Let's do that. using the `·` character so it is clear we didn't mean `.`

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 5 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1393]: Let's go for the clearer one.
(I need to review this page and split it up anyhow if we want it to be translatable - another few lines won't make any difference)

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 6 above)
Comment (MarkvanMents) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1649]: OK - should we add the note about it being database specific to encourage the user to test it?

**Review by MarkvanMents — COMMENTED** [review]
Thanks for your answers. I've had a quick look. 
Let me know if you have any questions, or would like me to draft any changes.
Mark

File: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md (hunk 6 above)
Comment (passalaqua) [reply content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md:1649]: `RPAD` is consistent. It's only `LPAD` that behaves differently in `HSQLDB` and other dbs.

**Comment by passalaqua** [comment]
For the record, both this and https://github.com/mendix/docs/pull/11250/ affect the same file and will need to be merged for 11.12.

**Review by MarkvanMents — APPROVED** [review]
Thanks for the PR. Ready for publication with 11.12 release.
